src/utils/stats_alloc/src/task_local.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// FIXME: This is a false-positive clippy test, remove this while bumping toolchain.
+// https://github.com/tokio-rs/tokio/issues/4836
+// https://github.com/rust-lang/rust-clippy/issues/8493
+#![expect(clippy::declare_interior_mutable_const)]
+
+use std::alloc::{GlobalAlloc, Layout, System};
+use std::future::Future;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::time::Duration;
+
+use tokio::task_local;
+
+#[repr(transparent)]
+#[derive(Clone, Copy)]
+struct TaskLocalBytesAllocated(Option<&'static AtomicUsize>);
+
+impl Default for TaskLocalBytesAllocated {
+    fn default() -> Self {
+        Self(Some(Box::leak(Box::new_in(0.into(), System))))
+    }
+}
+
+impl TaskLocalBytesAllocated {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub const fn invalid() -> Self {
+        Self(None)
+    }
+
+    #[inline(always)]
+    pub fn add(&self, val: usize) {
+        if let Some(bytes) = self.0 {
+            bytes.fetch_add(val, Ordering::Relaxed);
+        }
+    }
+
+    /// # Safety
+    /// The caller must ensure that `self` is valid.
+    #[inline(always)]
+    pub unsafe fn add_unchecked(&self, val: usize) {
+        self.0.unwrap_unchecked().fetch_add(val, Ordering::Relaxed);
+    }
+
+    #[inline(always)]
+    pub fn sub(&self, val: usize) {
+        if let Some(bytes) = self.0 {
+            let old_bytes = bytes.fetch_sub(val, Ordering::Relaxed);
+            if old_bytes - val == 0 {
+                unsafe { Box::from_raw_in(bytes.as_mut_ptr(), System) };
+            }
```

Comment:
We're actually emulating an `Arc`, and the ordering here might be problematic. Need further investigation.

https://doc.rust-lang.org/src/alloc/sync.rs.html#1664

Reply:
If there're adding or subtracting `0` bytes, this doesn't work as well.

There's no guarantee that the size of layout is non-zero. We may need to check this explicitly. Alternative: take the 8 byte of counter into account with `wrapped_layout.size()`.

> This function is unsafe because undefined behavior can result if the caller does not ensure that layout has non-zero size.